File: backend/api/schemas/investigation.py
from datetime import datetime

from pydantic import BaseModel


class InvestigationNoteBase(BaseModel):
    content: str

class InvestigationNoteResponse(InvestigationNoteBase):
    id: str
    alert_id: str
    analyst_username: str
    created_at: datetime
    model_config = {"from_attributes": True}

class InvestigationNoteCreate(InvestigationNoteBase):
    pass


class ChecklistItemResponse(BaseModel):
    id: str
    alert_id: str
    description: str
    is_checked: bool
    checked_by: str | None = None
    ai_rationale: str | None = None
    sort_order: int
    model_config = {"from_attributes": True}


class ChatMessageBase(BaseModel):
    content: str

class ChatMessageResponse(BaseModel):
    id: str
    alert_id: str
    role: str
    content: str
    analyst_username: str | None = None
    created_at: datetime
    model_config = {"from_attributes": True}

class ChatRequest(ChatMessageBase):
    pass


class SARDraftResponse(BaseModel):
    id: str
    alert_id: str
    version: int
    subject_info: str | None = None
    activity_description: str | None = None
    narrative: str | None = None
    reason_for_suspicion: str | None = None
    action_taken: str | None = None
    generated_by: str
    created_at: datetime
    model_config = {"from_attributes": True}

class SARDraftUpdate(BaseModel):
    subject_info: str | None = None
    activity_description: str | None = None
    narrative: str | None = None
    reason_for_suspicion: str | None = None
    action_taken: str | None = None


class AuditTrailEntryResponse(BaseModel):
    id: str
    alert_id: str
    action: str
    details: str | None = None
    performed_by: str
    created_at: datetime
    model_config = {"from_attributes": True}


class SimilarCaseResponse(BaseModel):
    """A case similar to the target alert, with a computed similarity score."""

    id: str
    alert_id: str
    title: str
    typology: str
    risk_score: int
    status: str
    resolution: str | None = None
    similarity_score: int
    matching_factors: list[str]
    model_config = {"from_attributes": True}

File: frontend/src/types/investigation.ts
export interface InvestigationNote {
  id: string;
  alert_id: string;
  analyst_username: string;
  content: string;
  created_at: string;
}

export interface ChecklistItem {
  id: string;
  alert_id: string;
  description: string;
  is_checked: boolean;
  checked_by: string | null;
  ai_rationale: string | null;
  sort_order: number;
}

export interface ChatMessage {
  id: string;
  alert_id: string;
  role: string;
  content: string;
  analyst_username: string | null;
  created_at: string;
}

export interface SARDraft {
  id: string;
  alert_id: string;
  version: number;
  subject_info: string | null;
  activity_description: string | null;
  narrative: string | null;
  reason_for_suspicion: string | null;
  action_taken: string | null;
  generated_by: string;
  created_at: string;
}

export interface AuditTrailEntry {
  id: string;
  alert_id: string;
  action: string;
  details: string | null;
  performed_by: string;
  created_at: string;
}

export interface NetworkNode {
  id: string;
  label: string;
  type: 'customer' | 'account' | 'counterparty';
  risk: string | null;
}

export interface NetworkEdge {
  source: string;
  target: string;
  amount: number;
  type: string;
  date: string;
  direction: string;
  counterparty: string;
}

export interface NetworkGraph {
  nodes: NetworkNode[];
  edges: NetworkEdge[];
}

export interface SimilarCase {
  id: string;
  alert_id: string;
  title: string;
  typology: string;
  risk_score: number;
  status: string;
  resolution: string | null;
  similarity_score: number;
  matching_factors: string[];
}
